File: fingerprint/src/main/java/com/alexnikola/fingerprint/FingerPrintDialog.java
package com.alexnikola.fingerprint;

import android.content.Context;
import android.hardware.fingerprint.FingerprintManager;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import org.jetbrains.annotations.Nullable;

public class FingerPrintDialog extends DialogFragment
        implements TextView.OnEditorActionListener, UiHelper.Callback {

    private static final String ARG_PASSWORD = "ARG_PASSWORD";
    private static final String ARG_WITH_FINGER = "ARG_WITH_FINGER";
    private static final String ARG_TITLE = "ARG_TITLE";
    private static final String ARG_FINGER_MESSAGE = "ARG_FINGER_MESSAGE";
    private static final String ARG_PASSWORD_MESSAGE = "ARG_PASSWORD_MESSAGE";

    @Nullable
    private String password;

    private boolean withFinger = true;

    @Nullable
    private String title;

    @Nullable
    private String fingerMessage;

    @Nullable
    private String passwordMessage;

    private Button cancelButton;
    private Button secondDialogButton;
    private View fingerprintContent;
    private View backupContent;
    private EditText passwordEditText;
    private TextView passwordDescriptionTextView;
    private TextView newFingerprintEnrolledTextView;

    private Stage stage = Stage.FINGERPRINT;

    private FingerprintManager.CryptoObject cryptoObject;
    private UiHelper uiHelper;

    private InputMethodManager inputMethodManager;

    @Nullable
    private Callback callback;

    public static FingerPrintDialog newInstance(@Nullable String password,
                                                boolean withFinger,
                                                @Nullable String title,
                                                @Nullable String fingerMessage,
                                                @Nullable String passwordMessage) {

        FingerPrintDialog dialog = new FingerPrintDialog();
        Bundle args = new Bundle();
        args.putString(ARG_PASSWORD, password);
        args.putBoolean(ARG_WITH_FINGER, withFinger);
        args.putString(ARG_TITLE, title);
        args.putString(ARG_FINGER_MESSAGE, fingerMessage);
        args.putString(ARG_PASSWORD_MESSAGE, passwordMessage);
        dialog.setArguments(args);
        return dialog;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof Callback) {
            callback = (Callback) context;
        } else if (getParentFragment() instanceof Callback) {
            callback = (Callback) getParentFragment();
        }

        inputMethodManager = (InputMethodManager) context
                .getSystemService(Context.INPUT_METHOD_SERVICE);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        setStyle(DialogFragment.STYLE_NORMAL, android.R.style.Theme_Material_Light_Dialog);

        if (getArguments() != null) {
            password = getArguments().getString(ARG_PASSWORD);
            withFinger = getArguments().getBoolean(ARG_WITH_FINGER);
            title = getArguments().getString(ARG_TITLE);
            fingerMessage = getArguments().getString(ARG_FINGER_MESSAGE);
            passwordMessage = getArguments().getString(ARG_PASSWORD_MESSAGE);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        if (title != null) {
            getDialog().setTitle(title);
        }

        View view = inflater.inflate(R.layout.fingerprint_dialog_container, container, false);

        FingerprintManager fingerprintManager = null;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            fingerprintManager = getActivity().getSystemService(FingerprintManager.class);
        }

        uiHelper = new UiHelper(
                fingerprintManager,
                (ImageView) view.findViewById(R.id.fingerprintIcon),
                (TextView) view.findViewById(R.id.fingerprintStatus), this);

        view.findViewById(R.id.fingerprintIcon);
                view.findViewById(R.id.fingerprintStatus);

        cancelButton = view.findViewById(R.id.cancelButton);
        secondDialogButton = view.findViewById(R.id.secondDialogButton);

        fingerprintContent = view.findViewById(R.id.fingerprintContainer);
        backupContent = view.findViewById(R.id.backupContainer);

        passwordEditText = view.findViewById(R.id.passwordEditText);
        passwordDescriptionTextView = view.findViewById(R.id.passwordDescriptionTextView);

        newFingerprintEnrolledTextView = view.findViewById(R.id.newFingerprintEnrolledTextView);




        cancelButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
            }
        });

        secondDialogButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (stage == Stage.FINGERPRINT) {
                    goToBackup();
                } else {
                    verifyPassword();
                }
            }
        });


        passwordEditText.setOnEditorActionListener(this);

        updateStage();

        // If fingerprint authentication is not available, switch immediately to the backup
        // (password) screen.
        if (!uiHelper.isFingerprintAuthAvailable()) {
            goToBackup();
        }
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        if (stage == Stage.FINGERPRINT) {
            uiHelper.startListening(cryptoObject);
        }
    }

    public void setStage(Stage stage) {
        this.stage = stage;
    }

    @Override
    public void onPause() {
        super.onPause();
        uiHelper.stopListening();
    }



    /**
     * Sets the crypto object to be passed in when authenticating with fingerprint.
     */
    public void setCryptoObject(FingerprintManager.CryptoObject cryptoObject) {
        this.cryptoObject = cryptoObject;

    }

    /**
     * Switches to backup (password) screen. This either can happen when fingerprint is not
     * available or the user chooses to use the password authentication method by pressing the
     * button. This can also happen when the user had too many fingerprint attempts.
     */
    private void goToBackup() {
        stage = Stage.PASSWORD;
        updateStage();
        passwordEditText.requestFocus();

        // Show the keyboard.
        passwordEditText.postDelayed(mShowKeyboardRunnable, 500);

        // Fingerprint is not used anymore. Stop listening for it.
        uiHelper.stopListening();
    }

    /**
     * Checks whether the current entered password is correct, and dismisses the the dialog and
     * let's the activity know about the result.
     */
    private void verifyPassword() {
        if (!checkPassword(passwordEditText.getText().toString())) {
            return;
        }
        /*if (stage == Stage.NEW_FINGERPRINT_ENROLLED) {
            SharedPreferences.Editor editor = mSharedPreferences.edit();
            editor.putBoolean(getString(R.string.use_fingerprint_to_authenticate_key),
                    mUseFingerprintFutureCheckBox.isChecked());
            editor.apply();

            if (mUseFingerprintFutureCheckBox.isChecked()) {
                // Re-create the key so that fingerprints including new ones are validated.
                mActivity.createKey(MainActivity.DEFAULT_KEY_NAME, true);
                stage = Stage.FINGERPRINT;
            }
        }
        passwordEditText.setText("");
        mActivity.onPurchased(false  without Fingerprint , null);*/
        dismiss();
    }

    /**
     * @return true if {@code password} is correct, false otherwise
     */
    private boolean checkPassword(String password) {
        // Assume the password is always correct.
        // In the real world situation, the password needs to be verified in the server side.
        return password.length() > 0;
    }

    private final Runnable mShowKeyboardRunnable = new Runnable() {
        @Override
        public void run() {
            inputMethodManager.showSoftInput(passwordEditText, 0);
        }
    };

    private void updateStage() {
        switch (stage) {
            case FINGERPRINT:
                cancelButton.setText(R.string.cancel);
                secondDialogButton.setText(R.string.use_password);
                fingerprintContent.setVisibility(View.VISIBLE);
                backupContent.setVisibility(View.GONE);
                break;
            case NEW_FINGERPRINT_ENROLLED:
                // Intentional fall through
            case PASSWORD:
                cancelButton.setText(R.string.cancel);
                secondDialogButton.setText(R.string.ok);
                fingerprintContent.setVisibility(View.GONE);
                backupContent.setVisibility(View.VISIBLE);
                if (stage == Stage.NEW_FINGERPRINT_ENROLLED) {
                    passwordDescriptionTextView.setVisibility(View.GONE);
                    newFingerprintEnrolledTextView.setVisibility(View.VISIBLE);
                }
                break;
        }
    }

    @Override
    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
        if (actionId == EditorInfo.IME_ACTION_GO) {
            verifyPassword();
            return true;
        }
        return false;
    }

    @Override
    public void onAuthenticated() {
        // Callback from UiHelper. Let the activity know that authentication was
        // successful.
        //mActivity.onPurchased(true /* withFingerprint */, cryptoObject);
        dismiss();
    }

    @Override
    public void onError() {
        goToBackup();
    }



    /**
     * Enumeration to indicate which authentication method the user is trying to authenticate with.
     */
    public enum Stage {
        FINGERPRINT,
        NEW_FINGERPRINT_ENROLLED,
        PASSWORD
    }

    public interface Callback {

    }
}

File: fingerprint/src/main/java/com/alexnikola/fingerprint/FingerPrintDialog.kt
package com.alexnikola.fingerprint

import android.content.Context
import android.hardware.fingerprint.FingerprintManager
import android.os.Build
import android.os.Bundle
import android.support.annotation.RequiresApi
import android.support.v4.app.DialogFragment
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import kotlinx.android.synthetic.main.fingerprint_dialog_backup.*
import kotlinx.android.synthetic.main.fingerprint_dialog_backup.view.*
import kotlinx.android.synthetic.main.fingerprint_dialog_container.*
import kotlinx.android.synthetic.main.fingerprint_dialog_container.view.*
import kotlinx.android.synthetic.main.fingerprint_dialog_content.*
import kotlinx.android.synthetic.main.fingerprint_dialog_content.view.*

@RequiresApi(api = Build.VERSION_CODES.M)
class FingerPrintDialog : DialogFragment(), TextView.OnEditorActionListener, FingerHelper.Callback {

    private var stage = Stage.FINGERPRINT

    private var password: String? = null
    private var withFinger = true
    private lateinit var title: String
    private lateinit var fingerMessage: String
    private lateinit var passwordMessage: String

    private var cryptoObject: FingerprintManager.CryptoObject? = null
    private var callback: Callback? = null

    private lateinit var fingerHelper: FingerHelper
    private lateinit var inputMethodManager: InputMethodManager

    private val mShowKeyboardRunnable = Runnable { inputMethodManager.showSoftInput(passwordEditText, 0) }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is Callback) {
            callback = context
        } else if (parentFragment is Callback) {
            callback = parentFragment as Callback
        }
        inputMethodManager = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
        setStyle(DialogFragment.STYLE_NORMAL, android.R.style.Theme_Material_Light_Dialog)

        withFinger = arguments?.getBoolean(ARG_WITH_FINGER) ?: false
        password = arguments?.getString(ARG_PASSWORD)
        title = arguments?.getString(ARG_TITLE) ?: getString(R.string.title)
        fingerMessage = arguments?.getString(ARG_FINGER_MESSAGE) ?: getString(R.string.fingerprint_description)
        passwordMessage = arguments?.getString(ARG_PASSWORD_MESSAGE) ?: getString(R.string.password_description)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        dialog.setTitle(title)

        val view = inflater.inflate(R.layout.fingerprint_dialog_container, container, false)

        fingerHelper = FingerHelper(activity, view.fingerprintIcon, view.fingerprintStatus, this)

        view.cancelButton.setOnClickListener { dismiss() }

        view.secondDialogButton.setOnClickListener {
            if (stage == Stage.FINGERPRINT) {
                goToBackup()
            } else {
                verifyPassword()
            }
        }
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.passwordEditText.setOnEditorActionListener(this)
        updateStage()

        // If fingerprint authentication is not available, switch immediately to the backup
        // (password) screen.
        if (!fingerHelper.isFingerprintAuthAvailable) {
            goToBackup()
        }
    }

    override fun onResume() {
        super.onResume()
        if (stage == Stage.FINGERPRINT) {
            fingerHelper.startListening(cryptoObject)
        }
    }

    fun setStage(stage: Stage) {
        this.stage = stage
    }

    override fun onPause() {
        super.onPause()
        fingerHelper.stopListening()
    }


    /**
     * Sets the crypto object to be passed in when authenticating with fingerprint.
     */
    fun setCryptoObject(cryptoObject: FingerprintManager.CryptoObject) {
        this.cryptoObject = cryptoObject

    }

    /**
     * Switches to backup (password) screen. This either can happen when fingerprint is not
     * available or the user chooses to use the password authentication method by pressing the
     * button. This can also happen when the user had too many fingerprint attempts.
     */
    private fun goToBackup() {
        stage = Stage.PASSWORD
        updateStage()
        view!!.passwordEditText.requestFocus()

        // Show the keyboard.
        view!!.passwordEditText.postDelayed(mShowKeyboardRunnable, 500)

        // Fingerprint is not used anymore. Stop listening for it.
        fingerHelper.stopListening()
    }

    /**
     * Checks whether the current entered password is correct, and dismisses the the dialog and
     * let's the activity know about the result.
     */
    private fun verifyPassword() {
        if (!checkPassword(passwordEditText!!.text.toString())) {
            return
        }
        /*if (stage == Stage.NEW_FINGERPRINT_ENROLLED) {
            SharedPreferences.Editor editor = mSharedPreferences.edit();
            editor.putBoolean(getString(R.string.use_fingerprint_to_authenticate_key),
                    mUseFingerprintFutureCheckBox.isChecked());
            editor.apply();

            if (mUseFingerprintFutureCheckBox.isChecked()) {
                // Re-create the key so that fingerprints including new ones are validated.
                mActivity.createKey(MainActivity.DEFAULT_KEY_NAME, true);
                stage = Stage.FINGERPRINT;
            }
        }
        passwordEditText.setText("");
        mActivity.onPurchased(false  without Fingerprint , null);*/
        dismiss()
    }

    /**
     * @return true if `password` is correct, false otherwise
     */
    private fun checkPassword(password: String): Boolean {
        // Assume the password is always correct.
        // In the real world situation, the password needs to be verified in the server side.
        return password.length > 0
    }

    private fun updateStage() {
        when (stage) {
            FingerPrintDialog.Stage.FINGERPRINT -> {
                view!!.cancelButton.setText(R.string.cancel)
                view!!.secondDialogButton.setText(R.string.use_password)
                view!!.fingerprintContainer.visibility = View.VISIBLE
                view!!.backupContainer.visibility = View.GONE
            }
            FingerPrintDialog.Stage.NEW_FINGERPRINT_ENROLLED,
                // Intentional fall through
            FingerPrintDialog.Stage.PASSWORD -> {
                view!!.cancelButton.setText(R.string.cancel)
                view!!.secondDialogButton.setText(R.string.ok)
                view!!.fingerprintContainer.visibility = View.GONE
                view!!.backupContainer.visibility = View.VISIBLE
                if (stage == Stage.NEW_FINGERPRINT_ENROLLED) {
                    view!!.passwordDescriptionTextView.visibility = View.GONE
                    view!!.newFingerprintEnrolledTextView.visibility = View.VISIBLE
                }
            }
        }
    }

    override fun onEditorAction(v: TextView, actionId: Int, event: KeyEvent): Boolean {
        if (actionId == EditorInfo.IME_ACTION_GO) {
            verifyPassword()
            return true
        }
        return false
    }

    override fun onAuthenticated() {
        // Callback from FingerHelper. Let the activity know that authentication was
        // successful.
        //mActivity.onPurchased(true /* withFingerprint */, cryptoObject);
        dismiss()
    }

    override fun onError() {
        goToBackup()
    }


    /**
     * Enumeration to indicate which authentication method the user is trying to authenticate with.
     */
    enum class Stage {
        FINGERPRINT,
        NEW_FINGERPRINT_ENROLLED,
        PASSWORD
    }

    interface Callback

    companion object {

        private val ARG_PASSWORD = "ARG_PASSWORD"
        private val ARG_WITH_FINGER = "ARG_WITH_FINGER"
        private val ARG_TITLE = "ARG_TITLE"
        private val ARG_FINGER_MESSAGE = "ARG_FINGER_MESSAGE"
        private val ARG_PASSWORD_MESSAGE = "ARG_PASSWORD_MESSAGE"

        @RequiresApi(api = Build.VERSION_CODES.M)
        fun newInstance(password: String?,
                        withFinger: Boolean,
                        title: String?,
                        fingerMessage: String?,
                        passwordMessage: String?): FingerPrintDialog {

            val dialog = FingerPrintDialog()
            val args = Bundle()
            args.putString(ARG_PASSWORD, password)
            args.putBoolean(ARG_WITH_FINGER, withFinger)
            args.putString(ARG_TITLE, title)
            args.putString(ARG_FINGER_MESSAGE, fingerMessage)
            args.putString(ARG_PASSWORD_MESSAGE, passwordMessage)
            dialog.arguments = args
            return dialog
        }
    }
}